Laravel Upgrade: Fix Unresolved Issues

ardelato opened the pull request:
## Description

This is a hefty pull of changes.

The first few commits eb96bd020f42d48af47703a2ff1bf3a8d50f7630 - 0f1d6162f92bfcf1c3843f0ec04192626d2cbe99 revolve around configuring Docker and local dev to be able to run the suite of PHP tests in the repo.

The later commits revolve around fixing issues from the Laravel upgrades (10, 11, to 12). We were unable to catch these at the time because we were not able to run tests.

### CR Notes

There is a mix of changes revolving updating the tests and testing setup and actually addressing the issues for test failures.

The changes to the testing setup were needed to speed up the tests and thus the feedback loop. Most tests were taken 1min+ each so the fast login method and usage of transaction setup/teardown allowed said tests to take about 10s.

One thing I do wish to do in the future is migrate to [Laravel Sanctum](https://laravel.com/docs/12.x/sanctum) for the authentication system. At the moment, there is a very janky middleware handling the management and retrieval of user tokens.

Additionally, I skipped some tests as the functionality being tested is surrounding services and features we won't be using - i.e. Discourse, Wordpress, MediaWiki.

qa_req 0 All tests should pass, which have to be exec'd locally.

ardelato committed:
fix: remove quotes from CLI_ARGS in Taskfile.yml

When using the docker:run:artisan task with additional arguments,
the quoted CLI_ARGS parameter prevents proper argument parsing.

For example, running:

  task docker:run:artisan -- test --filter=SomeTest

would fail because the quotes around CLI_ARGS concatenate all arguments as a
single string rather than passing them individually to the artisan command.

Removing these quotes allows proper argument passing to artisan commands with
multiple parameters.

ardelato committed:
feat(taskfile): add docker:run:tests task

ardelato committed:
feat(env): implement custom environment loader for .env.testing

This introduces a custom environment variable loader to address a
limitation in Laravel's default .env handling. We will load both
environment-specific files (like .env.testing) and the base .env file,
allowing specific environments to override only the variables they need
without duplicating all configuration. Without this, all env variables
not declared in .env.testing would be missing.

By doing all this, we can now extract some of the env variables from
the phpunit.xml file. The end goal is to utilize the same MySQL container
for both dev and testing but obviously we'd use different databases.
This would simplify things as we would not need to spin up a separate
MySQL container for testing.

ardelato committed:
fix(middleware): re-fix the APISetLocale middleware

Laravel Shift had re-added type hints that break APISetLocale.php.
I needed to reapply the fix from
https://github.com/iFixit/restarters/commit/3e6006b73626199c38c4c0c637bf65b992909b4b

ardelato committed:
fix(openapi): generate the OpenAPI docs on startup

We run php artisan l5-swagger:generate in CircleCI and it is needed to
run our tests. Otherwise we get the following error:

cebe\openapi\SpecBaseObject::__construct(): Argument #1 ($data) must be
of type array, null given, called in /var/www/vendor/devizzent/cebe-php-openapi/src/Reader.php on line 39

As such, we will now generate the OpenAPI docs on startup and ignore
the storage/api-docs directory.

ardelato committed:
feat(taskfile): add task to initialize the test database

We will need to run this task before running the tests, but we would be
able to utilize the same MySQL container for tests.

We will pull in the .env.testing file to override the .env.docker vars
that point to the dev db. That said, there seems to be a current bug
with Task where task specific dotenvs values are not accessible with the
Go template engine, while the global dotenv values are. As such,
we need to use the $VAR_NAME syntax to access the task specific dotenv
values.

ardelato committed:
patch(task): drop .taskrc.yml with new Task version

The MAP_VARIABLES experimental feature has now been added to newer
Task versions, so we can drop the .taskrc.yml file.

ardelato committed:
fix(phpunit): run phpunit migration configuration

We've updated our phpunit version over the past Laravel upgrades but
have not updated our testing stuff because we were unable to run
tests locally. Now that we are able to run tests locally, we can
begin to address all the issues surrounding the testing stuff.

ardelato committed:
refactor(tests): update test annotations to use `#[Test]` attribute

Replaced the deprecated PHPDoc style test annotations with the new `#[Test]`
attribute in various test files to improve compatibility with PHPUnit 11 and above.

ardelato committed:
refactor(tests): update test annotations to use `#[DataProvider]` attribute

Replaced deprecated PHPDoc style data provider annotations with the new `#[DataProvider]` attribute across various test files to enhance compatibility with PHPUnit 11 and improve code clarity.

ardelato committed:
chore(tests): remove obsolete test files

Deleted empty test files EventKeyTest.php and GroupByKeyTest.php to clean up warnings

ardelato committed:
refactor(tests): rename DiscourseTest to SwaggerGenerateTest

Updated the test class name to SwaggerGenerateTest to adhere to PSR
standards and cleanup PHPUnit warnings.

ardelato committed:
refactor(tests): change test provider methods to static

Updated various test classes to define data provider methods as static to address PHPUnit warnings.

ardelato committed:
chore(tests): update .gitignore to include PHPUnit cache files

Added .phpunit.result.cache and .phpunit.cache/ to .gitignore.

ardelato committed:
refactor(user): change return type of lastLogin method to Carbon instance

Updated the lastLogin method in the User model to return a Carbon instance instead of a Date. This was causing tests to fail.

ardelato committed:
chore(tests): configure testing environment and bootstrap file

Updated the testing environment by adding APP_ENV to .env.testing,
modified phpunit.xml to use tests/bootstrap.php, and adjusted the
Taskfile to run tests with the testing environment. A new bootstrap
file was created to load necessary dependencies for tests if we run
phpunit directly.

ardelato committed:
refactor(tests): optimize test setup and teardown process

Trimmed the test setup by truncating only necessary tables and using
transactions for faster execution. Enhanced OpenAPI validation checks
to only run when applicable.

ardelato committed:
feat(tests): add ApiTestCase for streamlined API testing

Introduced a new base test case class, ApiTestCase, to simplify API
testing by handling authentication and providing helper methods for
HTTP requests with API tokens. Updated AlertsTest to extend ApiTestCase
and utilize the new fast login method for improved test performance.

ardelato committed:
fix(events): add __invoke method to BaseEvent

This issue was found while running tests.

The core issue is that Laravel 12 changed the way queued events are handled.
In Laravel 12, the queue system expects listener classes to implement
an __invoke() method when processing queued events, which is a change
from earlier versions where handle() was called directly.

The error "Call to undefined method __invoke()" was occurring because
our BaseEvent class and its children didn't implement this method.

This was probably missed during our Laravel 9-12 upgrades since we were
not able to run tests locally.

ardelato committed:
fix(observer): resolve Event and User model type errors in EventsUsersObserver

This commit fixes type checking errors in the EventsUsersObserver class that appeared
after upgrading to Laravel 12 and only noticed when running tests.

1. Added proper App\Models\Party import to fix "must be of type App\Observers\Party" error
2. Changed model retrieval to use relationships and where().first() to ensure single
   model instances instead of collections
3. Updated method signatures with nullable type hints (?Party, ?User) to handle
   potential null values
4. Added null checks before accessing properties to prevent errors on null objects
5. Fixed the removed() method signature to properly handle the wasConfirmed parameter

These changes address stricter type checking in PHP 8.2 and Laravel 12's
updated Eloquent behavior for model resolution.

ardelato committed:
refactor: dedupe code in EventsUsersObserver

ardelato committed:
refactor(tests): skip Discourse tests when feature is disabled

We won't be running Discourse so there is no need to run the related tests.
As such, let's extract some of the current feature flag check into a base test case
to make it easier to skip the tests when the feature is disabled.

ardelato committed:
fix(models): address SQL syntax errors in geographic queries

Improved SQL query parameter binding in latitude/longitude calculations
to prevent syntax errors with floating point numbers. Changes include:

- User::groupsNearby() and nearbyRestarters(): replaced string interpolation
  with parameterized queries using selectRaw() and parameter binding
- Party::upcomingEventsInUserArea(): implemented parameter binding for
  coordinates in the haversine distance calculation

ardelato committed:
fix(tests): update BasicTests to be more resilient to database changes

Refactored tests to handle ID sequence variations after Laravel upgrade
and test optimization changes:

- Groups/BasicTest.php: Modified assertions to find the specific group
  being tested in the returned list instead of assuming array position
- Fixometer/BasicTest.php: Changed assertions to verify data structure
  and field presence rather than exact ID values
- Removed strict network assertions that were assuming specific data ordering

These changes make tests more resilient to environmental differences,
database state variations, and the recent change from truncation to
transaction-based test cleanup.

ardelato committed:
refactor(tests): optimize BasicTest classes for faster test execution

Improved test performance and reliability across multiple BasicTest files:

- Dashboard/BasicTest:
  - Added fastLoginAsTestUser to avoid HTTP requests
  - Reduced queue processing to a single operation
  - Added direct attribute assignments for entity creation

- Fixometer/BasicTest:
  - Used more efficient file operations with proper checks
  - Cleaned up test data creation with minimal attributes

- Notifications/BasicTest:
  - Fixed notification handling to avoid type errors
  - Set required language field for locale management
  - Used real notifications instead of fakes for DB verification

These changes significantly decrease test execution time from ~60s to sub 5s
per test.

ardelato committed:
fix(tests): update CategoryTest classes to extend ApiTestCase

Similar error as the AlertsTest where the api calls returned an
Unauthenticated error. Let's use the ApiTestCase for these tests.

Let's also fix the namespace to match their correct location in the codebase
and use the fastLoginAsTestUser method to speed up the tests.

ardelato committed:
fix(sql): properly handle env variables in Device::ofThisEvent query

Use parameter binding for environment variables in SQL instead of
attempting to inject them directly as string literals. This fixes
SQL syntax errors when querying devices for an event.

The issue was exposed during the Laravel 12 upgrade which has stricter
SQL query handling compared to Laravel 9.

ardelato committed:
test(groups): fix database cleanup in GroupCreateTest

Add proper database cleanup in GroupCreateTest after we've moved to
transaction-based TestCase setup.

ardelato committed:
fix(api): update Group model scopes and controller for Laravel 12

The scope prefix was chaned to #[Scope] attributes during our Laravel
upgrades. However, the scope of the methods were not updated properly
and as such broke the API.

1. Methods are now public with the attribute to be picked up by the
   API.
2. Reimplements the moderateGroupsv2 method in GroupController to use
   proper query builder patterns instead of attempting to call the scope
   method statically, which is not supported in Laravel 12.

These changes fix API errors related to scope method calls and ensure
proper handling of group moderation queries for administrators and
network coordinators.

ardelato committed:
feat(tests): add helper function for API test authentication

This adds a helper function to create a user with a unique API token.
This is similar to what we did in GroupCreateTest but we would need to
do this for all tests that used a static and non-unique token. This
all stems from switching to a transaction-based test setup and tests
making API calls which bypass the transaction event system -- can't
undo DB transactions if queries are made outside the testing context.

As such, this is a helper function that can be used in all tests that
need to create a specific user role for API tests and ensures that
the token won't cause conflicts.

ardelato committed:
fix(tests): use unique group names in APIv2GroupTest

Updated the APIv2GroupTest to generate unique group names using uniqid()
for all group creation and validation scenarios.

ardelato committed:
fix(image): enhance image processing for testing mode in FixometerFile

Updated the FixometerFile class to skip image manipulations when in test mode,
avoiding dependency issues. Added logic to create empty thumbnails during testing
and ensured the uploads directory is created if it doesn't exist.

This maybe a temporary fix until we investigate if the intervention-image
package update was the cause of the issue.

ardelato committed:
fix(forms): implement Form facade

During the upgrade to Laravel 12, we switched from laravelcollective/html to
spatie/laravel-html. This resulted in a 'Class "Form" not found' error,
so we needed to manually implement a Form facade.

That said, we can probably remove Spatie's HTML package and use Laravel's
native HTML generation instead.

ardelato committed:
fix(tests): update AddRemoveVolunteerTest to use new helper methods

This should help resolve the api token taken issue and speed up
the test.

ardelato committed:
fix(models): update membersRestarters method to handle null query parameter

The membersRestarters method in the Group model now accepts an optional query parameter
to support both direct method calls and query builder chaining.

ardelato committed:
fix(docker): update session configuration fo Docker

The authentication errors when creating events or registering users might
be caused by incompatibilities between Laravel 12's cookie handling and our
Docker environment.

Laravel 12 enforces stricter cookie security policies, requiring explicit
configuration of SESSION_DOMAIN, SANCTUM_STATEFUL_DOMAINS, and SESSION_SECURE_COOKIE
to ensure proper authentication token storage and retrieval.

ardelato committed:
fix(tests): configure mailpit for tests

There are some tests that actually send emails, as such we need to
have mailhog running when running tests. That said, mailhog is no longer
maintained so we should switch to mailpit.

ardelato committed:
patch(middleware): add CustomApiTokenAuth middleware for api token auth

The middleware and routes were majorily modified during the Laravel upgrades.
The tests exposed there was an issue and I confirmed we were unable to
create users, events, and groups.

This is a temporary fix as we should really migrate over to Sanctum for
API authentication and authorization.

This middleware works by not rejecting the request immediately and
attempt to match the token against the following sources:
1. First checking if the user is already authenticated via session
2. Then attempting token validation via query params, request body, headers, and cookies
3. Finally authenticating the user and setting proper headers for downstream middleware

Without this middleware, API endpoints would reject requests with "unauthenticated" errors.

ardelato committed:
fix(tests): refactor APIv2EventTest to use role constants and improve assertions

Updated APIv2EventTest to utilize role constants for user creation, enhancing readability and maintainability.
Refined assertions for event moderation checks to ensure created events are correctly listed and unapproved.

ardelato committed:
fix(file handling): update file paths to use public_path for uploads

Refactored file handling in Fixometer and FixometerFile classes to utilize
public_path() instead of $_SERVER['DOCUMENT_ROOT'] for constructing upload
file paths. The prior expectation was that the uploads path was
/var/www/uploads, but it is now /var/www/public/uploads. As such,
we won't use $_SERVER['DOCUMENT_ROOT'] anymore.

ardelato committed:
fix(auth): enhance user authentication flow in GroupController

Refactored the getUser method to improve user authentication by checking
multiple methods: standard session, API token in query parameters,
Authorization header, and testing environment. Updated error handling
to throw an AuthenticationException with a clear message for
unauthenticated access.

Additionally, updated APIv2GroupTest to ensure proper testing of
authentication scenarios, including backward compatibility for
API token in query parameters.

ardelato committed:
fix(api): allow timezone endpoint to be accessed without authentication

The custom authentication method disallowed access to the timezone endpoint,
so we need to ensure that the endpoint is still accessible without authentication.

ardelato committed:
fix(controller): update return type for getGroupEmailsWithNames method

Changed the return type of the getGroupEmailsWithNames method in PartyController
to \Illuminate\Http\JsonResponse for better type hinting. Additionally, refactored
user creation in InviteEventTest to use createUserWithToken for improved readability.

ardelato committed:
fix(models): update User model to handle default locale configuration

Modified the getPreferredLocale method in the User model to return the user's
language or fallback to the application's default locale configuration. This change
ensures that a valid locale is always returned, improving localization support.

ardelato committed:
fix(tests): refactor user authentication in APIv2DeviceTest and APIv2NetworkTest

Updated test cases to use fastLoginAsTestUser and createUserWithToken for
user authentication.

ardelato committed:
fix(tests): update notification links to use dynamic domain configuration

Refactored notification test outputs to replace hardcoded URLs with dynamic
domain configuration. This change ensures that the links in email notifications
are correctly generated based on the application's URL, improving flexibility
and maintainability of the tests.

ardelato committed:
fix(models): enhance membersHosts method to support optional query parameter

Updated the membersHosts method in the Group model to accept an optional query parameter similar to memberRestarters.

ardelato committed:
fix(tests): truncate group_network table in GroupTest setup

Added truncation of the group_network table in the GroupTest class
setup method. This was a side effect of removing the truncation steps
in TestCase in favor of transactions.

ardelato committed:
feat(config): add WordPress integration feature toggle

We won't be using the Wordpress Integration so let's add a feature toggle
to disable it and thus skip the corresponding tests.

ardelato committed:
fix(models): correct SQL syntax in User model

I inccorectly modified this query when I was dropping the DB::raw calls.
This should fix the issue.

ardelato committed:
fix(controller): update UserController to return empty JSON response

Laravel Shift had set an explicit return type for the postEmail method,
but we have not updated the code to handle all paths. As such, lets
update the method to return an empty JSON response and update the test
assert this as opposed to just a null value.

ardelato committed:
fix(controllers): align UserController::edit with test expectations

The UserController::edit method was originally designed to handle GET requests,
after the type hinting was added by Laravel Shift, it broke the ProfileTest::testEdit.
It exposed the fact that the test was asserting behavior for POST requests to
this endpoint. This caused test failures as the response content didn't match
expectations.

This fix:
- Modifies the edit method to handle both GET and POST requests
- For POST requests, returns either an empty view for regular users/restarters,
  or the edit view for hosts/administrators, matching expected test behavior
- Properly initializes user groups data to avoid in_array() errors in the view
- Creates an empty view file to enable returning empty responses

The solution preserves the existing application behavior while ensuring tests
pass correctly, addressing a TODO comment in the test code that noted the
tested behavior seemed unusual but was how the system was coded.

The cleanest solution would be to create a new method to handle POST requests
specifically, but this would require a lot of changes to the codebase. For
now we will just patch the current method to work like before the type hinting
was added.

ardelato committed:
fix(tests): refactor ProfileTest to use dynamic API token generation

Updated the ProfileTest to create users with dynamic API tokens instead of hardcoding values.
This change improves test reliability and readability by ensuring that the token used in API
requests is always valid. Additionally, modified assertions to check for user presence in the
response rather than exact count, enhancing the robustness of the test.

ardelato committed:
fix(auth): update policy parameter handling for Users

The recent migration from Laravel 9 to Laravel 12 introduced breaking
changes in how policy authorization parameters are handled.

This fixes UserPolicy::changeRepairDirRole parameter handling:
- Updated method signature from (User $perpetrator, User $victim, int $role) to
     (User $user, User $victim, int $role) to follow Laravel 12's strict convention
- Modified Gate::authorize call to omit the authenticated user, as Laravel 12 now
     automatically injects this as the first parameter
- Added PHPDoc comments to clarify parameter roles and order

ardelato committed:
fix(tests): enhance MenusTest to include mediawiki username handling

Updated MenusTest to set a mediawiki username for the user before making
requests to the /user/menus endpoint. This change ensures that the test
validates the correct handling of the wiki_username parameter in the
request.

ardelato committed:
fix(tests): update GroupViewTest

Let's use the fastLogin method to speed up the tests and let's also use a different number than group id 1 when searching. There is a change the group id may already exist.

ardelato committed:
fix(tests): update GroupJoinTest discourse call assertion

Because we disabled the Discourse Integration in tests, the functionality
of the AddUserToDiscourseGroup listener was behaving differently than
we were expecting. As such, let's losen the assertion to not expect
a certain number of calls but instead expect at least one call.

ardelato committed:
chore(gitignore): ignore csvs generated from tests

ardelato commented:
dev_block ⚡ some tests started to fail. Probably from changes to fix some other tests that were previously failing.

mlahargou commented on .env.testing:
```diff
@@ -1,3 +1,5 @@
+APP_ENV=testing
```
Why here instead of adding these to the phpunit.xml config?

ardelato replied:
`APP_TESTING` specifically or all `.env` variables?

mlahargou replied:
I was referring to `APP_TESTING`.

ardelato replied:
Ah gotcha! 

This was because we'd also want to use this same `.env.testing` for the `artisan` commands which may not set `APP_ENV=testing`.

mlahargou commented on tests/Feature/Users/Registration/DiscourseAccountCreationTest.php:
```diff
@@ -14,10 +14,10 @@
 use Tests\Feature\MockInterface;
 
-class DiscourseAccountCreationTest extends TestCase
+class DiscourseAccountCreationTest extends DiscourseTestCase
```
A lot of these tests in this file used to be commented out. Let's just uncomment them and keep them around. We will already skip them b/c we disabled the feature.

That or just drop the tests that only say `assertTrue(true)` lol.

mlahargou replied:
And here: `tests/Feature/Users/Registration/DiscourseAccountDeletionTest.php`

ardelato replied:
I still don't know what we plan on doing with the whole Discourse service so I'll just remove the the `assertTrue` tests.

ardelato replied:
Addressed in [5f263cf](https://github.com/iFixit/restarters/pull/7/commits/5f263cf7cec809553a6be5fa56d27a8fabfd3c29)

ardelato committed:
fix(models): improve isStartingSoon method logic in Party model

Refined the isStartingSoon method to accurately determine if an event is
starting within the next 3 hours. We recently updated Carbon and the test
itself EventStateTest was failing which exposed the issue. The test itself
has stricter assertions about what is "starting soon" then the actual
method. The test asserts the event is within a certain time range, while
the method only check if the event is at least on the current day.

Let's update the method to check if the event is starting within the next
3 hours to align with the test's expectations.

mlahargou commented on tests/Feature/Groups/GroupCreateTest.php:
```diff
@@ -15,13 +15,33 @@
+        
         $user = User::factory()->administrator()->create([
-                                                                      'api_token' => '1234',
+                                                                      'api_token' => $uniqueToken,
```
Wild spacing here.

ardelato replied:
Yeah there are a lot of tests like this. I just didn't have the appetite to address them all so I addressed almost none 😅

ardelato commented:
un_dev_block ⚡ There was only one test that was failing, all other tests were flaky.

mlahargou commented on app/Http/Middleware/CustomApiTokenAuth.php:
```diff
@@ -0,0 +1,178 @@
+        $token = $request->query('api_token');
+        if (!empty($token)) {
+            $this->tokenSource = 'query';
+            Log::debug("Found token in query string", ['length' => strlen($token)]);
```
Do we want to keep all of these debug statements?

ardelato replied:
I was planning on removing it but I am going to come back to this file in the future to properly address the authentication problem and see about moving to Laravel Sanctum or Laravel Fortify. I could just re-add debug statements at that point in time but I might forget what I was trying to do originally.

mlahargou commented on app/Http/Middleware/CustomApiTokenAuth.php:
```diff
@@ -0,0 +1,178 @@
+    public function handle(Request $request, Closure $next)
+    {
+        // STEP 1: Log detailed request information
+        Log::debug("API Request", [
```
Is logging this some sort of security concern?

ardelato replied:
This would only show up if we set `LOG_LEVEL=debug` and even then it would get logged to a log file.

mlahargou commented on tests/Feature/Events/APIv2EventTest.php:
```diff
@@ -68,21 +74,30 @@ public function testGetEventsForGroup(): void {
+        $this->assertContains($id2, $eventIds, "Second created event should be in the moderation list");
+        
+        // Verify our events aren't approved
+        $ourEvents = collect($moderationEvents)->filter(function($event) use ($id1, $id2) {
```
I'm not familiar with the `collect` method.

ardelato replied:
collect would be similar to the early days of lodash and JS arrays, so it wraps an "array" and provides useful helper methods to interact with it.

mlahargou replied:
Okay 👍 

I think this could be
```
$ourEvents = array_filter(...);
```

And above, `collect($moderationEvents)->pluck('id')->toArray();` could be:

```
$eventIds = array_map(($event) => $event['id'], $moderationEvents);
```

ardelato replied:
Addressed in [28c7c8c](https://github.com/iFixit/restarters/pull/7/commits/28c7c8c3c7037f6ce6c18a489dbb8963fd99241f)

mlahargou commented on routes/api.php:
```diff
@@ -65,8 +65,9 @@
-Route::get('/timezones', [App\Http\Controllers\ApiController::class, 'timezones']);
+// Timezones - Publicly accessible endpoint
+Route::get('/timezones', [App\Http\Controllers\ApiController::class, 'timezones'])
+     ->withoutMiddleware(\App\Http\Middleware\CustomApiTokenAuth::class); // Explicitly bypass authentication
```
Cool!

mlahargou commented on tests/Feature/Networks/APIv2NetworkTest.php:
```diff
@@ -81,6 +83,10 @@ public function testGet(): void {
     #[DataProvider('providerGroupsParameters')]
     public function testListGroups($getNextEvent, $getDetails): void {
+        // Create a test user with admin privileges and authenticate
+        $user = $this->createUserWithToken(Role::ADMINISTRATOR);
+        $this->actingAs($user);
```
Should `createUserWithToken` just call `actingAs`? I guess maybe we just sometimes want the user and don't want them to be "logged in".

ardelato replied:
These tests do need to be audited. In some tests it would make sense to make actual HTTP requests to login, other tests we'd just want to have a user to use so the fast login method would be best, and then there are some tests where we just use the user api token for requests while other times we use the same user to do stuff.

The tests really need to be more divided than just directory paths.

mlahargou commented on tests/Feature/Events/WordpressEventPushTest.php:
```diff
@@ -27,6 +27,11 @@ protected function setUp(): void
         parent::setUp();
 
+        // Skip all tests if WordPress integration is disabled
+        if (config('restarters.features.wordpress_integration') === false) {
+            $this->markTestSkipped('WordPress integration is disabled.');
+        }
```
We should move this above `parent::setUp()`.

mlahargou replied:
Same in `tests/Feature/Groups/WordpressGroupPushTest.php`.

ardelato replied:
Ha so I just tried applying that change and looks like we do need it after the `parent::setUp` call. Otherwise, the `config` is not setup:

![image](https://github.com/user-attachments/assets/3d9b4731-6e36-4111-b363-7109323e28fb)

mlahargou replied:
Wild, okay 👍

mlahargou commented on tests/Feature/Users/Registration/AccountCreationTest.php:
```diff
@@ -146,7 +146,8 @@ public function testValidEmail(): void
 
-        $this->assertNull(json_decode($response->getContent(), true));
+        // Assert empty response.
+        $this->assertEquals([], json_decode($response->getContent(), true));
```
There's an `assertEmpty` in phpunit which would work here as well.

ardelato replied:
Addressed in [8b104c0](https://github.com/iFixit/restarters/pull/7/commits/8b104c0c998ca5cec39e9c82e4bcc9cc6f78d4c8)

mlahargou commented on app/Http/Controllers/UserController.php:
```diff
@@ -730,13 +730,73 @@ public function create(Request $request)
+            }
+        }
 
+        // Original GET request handling below
```
Drop this comment.

ardelato replied:
Addressed in [38ddedb](https://github.com/iFixit/restarters/pull/7/commits/38ddedb31af0c562b4340d9f4614f6d9b4062b95)

mlahargou commented on app/Models/Party.php:
```diff
@@ -450,19 +450,23 @@ public function isUpcoming()
+        if (!$this->isInProgress() && 
+            !$this->hasFinished() && 
+            $start->gt($now) && 
+            $now->diffInHours($start) <= 3) {
```
3 hours seems a bit too short. Why don't we update the test to be 23 hours out and this method to be within 24 hours.

ardelato replied:
Because I wasn't sure what the original intention was so I chose to use the test's expectation as the source of truth. This also makes more sense to me in respect to events. I wouldn't expect an event that is at 8pm to be "starting soon" if it was currently 6am.

mlahargou reviewed (approved):
CR 📱 With a few minor suggestions.

ardelato committed:
fix(tests): remove unneeded Discourse tests

ardelato committed:
refactor(tests): update AccountCreationTest assertion for empty response

Use assertEmpty instead of assertEquals empty array.

ardelato committed:
refactor(controllers): remove unnecessary comment

ardelato committed:
refactor(tests): simplify event ID extraction in APIv2EventTest

Use native PHP array functions instead of collect().

mlahargou reviewed (approved):
CR 📱
